### PR TITLE
[netatmo] Avoid endless loop when Security claims event history

### LINKED ISSUE #17478: [netatmo] Request storm leads to MAXIMUM_USAGE_REACHED
I identified that upon start my Netatmo system ramps-up very quick to ~ 65 requests in a short period of time. I guess this conflicts with the 50 requests / 10 seconds leading to `MAXIMUM_USAGE_REACHED`.

Having disabled everything, as soon as I enable a Security or Control child (welcome or plug in my case) requests starts to burst very quickly. This will have to be studied - the culprit seems to be around `getevents`: 

```
2024-09-28 14:11:30.385 [DEBUG] [l.handler.capability.CacheCapability] - HomeCapability requesting fresh data for netatmo:home:1:1
2024-09-28 14:11:30.386 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/homesdata?home_id=59959297b4809d1aff8b4726 
2024-09-28 14:11:30.540 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/homestatus?home_id=59959297b4809d1aff8b4726 
2024-09-28 14:11:30.647 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/homesdata?home_id=60be3f90fcb4a02d086ea475 
2024-09-28 14:11:30.884 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/homestatus?home_id=60be3f90fcb4a02d086ea475 
2024-09-28 14:11:31.222 [DEBUG] [handler.capability.RefreshCapability] - 'netatmo:home:1:1' next refresh in PT3M
2024-09-28 14:12:33.600 [DEBUG] [tatmo.internal.handler.DeviceHandler] - Initializing handler for bridge netatmo:plug:1:1
2024-09-28 14:12:35.594 [DEBUG] [er.capability.ParentUpdateCapability] - Requesting parents data update for Thing 'netatmo:plug:1:1'
2024-09-28 14:12:35.595 [DEBUG] [handler.capability.RefreshCapability] - 'netatmo:home:1:1' next refresh in PT2S
2024-09-28 14:12:35.601 [DEBUG] [er.capability.ParentUpdateCapability] - Requesting parents data update for Thing 'netatmo:plug:1:1'
2024-09-28 14:12:37.595 [DEBUG] [l.handler.capability.CacheCapability] - HomeCapability requesting fresh data for netatmo:home:1:1
2024-09-28 14:12:37.596 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/homestatus?home_id=59959297b4809d1aff8b4726 
2024-09-28 14:12:37.761 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/homestatus?home_id=60be3f90fcb4a02d086ea475 
2024-09-28 14:12:38.111 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475 
2024-09-28 14:12:38.207 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f7befcbdc02ebed30753be 
2024-09-28 14:12:38.309 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f72244c6b9944f320fb004 
2024-09-28 14:12:38.409 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f6cb6a79e82186ce0a65d2 
2024-09-28 14:12:38.538 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f69b9cae1bf9947d02f3ad 
2024-09-28 14:12:38.675 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f6779743fe0f788201295f 
2024-09-28 14:12:38.935 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f63aa0b485741fe508da2f 
2024-09-28 14:12:39.138 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f59be63d5ff3d343027aec 
2024-09-28 14:12:39.240 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f57edf8b0bc82d52068226 
2024-09-28 14:12:39.343 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f50a50357cc3a5ad0652ed 
2024-09-28 14:12:39.487 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f4ee3a79368dbe150090f9 
2024-09-28 14:12:39.763 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f4397f3aea0f786b00e576 
2024-09-28 14:12:40.032 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f3f8b604f21b25500f0744 
2024-09-28 14:12:40.270 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f3a16a50f29f26b60685d4 
2024-09-28 14:12:40.386 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f2d2edb4f936dfc7024e03 
2024-09-28 14:12:40.549 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f271e31e45b3da0b0567c9 
2024-09-28 14:12:40.763 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f1aa18993715681e07a5fb 
2024-09-28 14:12:40.914 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f168730c9aec1f7ed69765 
2024-09-28 14:12:41.071 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f10be0ad35233ab80ed57f 
2024-09-28 14:12:41.444 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:41.547 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:41.840 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:41.909 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:41.969 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:42.090 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:42.175 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:42.297 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:42.466 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:42.549 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:42.628 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:42.686 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:42.834 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:42.935 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:43.030 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:43.104 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:43.166 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:43.235 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:43.303 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:43.371 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:43.471 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:43.566 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:43.633 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:43.695 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:43.784 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:43.858 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:43.953 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:44.033 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:44.105 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:44.181 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:44.269 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:44.348 [DEBUG] [mo.internal.handler.ApiBridgeHandler] - executeUri GET  https://api.netatmo.com/api/getevents?home_id=60be3f90fcb4a02d086ea475&event_id=66f0f75e22137c82a60c26f2 
2024-09-28 14:12:44.405 [WARN ] [andler.capability.SecurityCapability] - Error retrieving last events for home '60be3f90fcb4a02d086ea475' : Rest call failed: statusCode=MAXIMUM_USAGE_REACHED, message=User usage reached
```

Here we can observe that event `66f0f75e22137c82a60c26f2` is requested 32 times before leading to `MAXIMUM_USAGE_REACHED`

### PR DESCRIPTION
**Pull request by clinique**
Resolves #17478 

While digging in this issue, I landed in the immediate MAXIMUM_USAGE_REACHED issue (see comment [here](https://community.openhab.org/t/netatmo-bridge-offline-99-of-time/149168/44?u=glhopital)

### CODE REVIEW

**Comment by clinique** [comment]
@lolodomo , @jlaur  : could you please have a quick look at this one ? I saw this morning that #17484 was onboarded in 4.2.2 - this will generate the need for a rapid fix.

**Comment by clinique** [comment]
Sorry, I meant issue #17478 , generated by PR #17448.
Removing the `@Nullable freshestEventTime` exposed an endless loop. The bug was somehow there before but not visible.

**Comment by lsiepel** [comment]
Could you give a summmary what causes this issue and what you did to fix it?

**Comment by clinique** [comment]
> Could you give a summmary what causes this issue and what you did to fix it?  

The default value choosen is never reached and calls to getevents had no other exit condition - creating an endless loop

**Comment by clinique** [comment]
Hello @lsiepel , is my comment enough for you to help progress on this PR?

**Comment by lsiepel** [comment]
Will look at it this weekend. How critical is this"issue, is the binding effectively non-useable in 4.2.2. now?

**Comment by clinique** [comment]
Yes. As soon as Security things goes online.

**Comment by lolodomo** [comment]
I am reviewing it now...

**Comment by clinique** [comment]
> I am reviewing it now...

I'm available

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java [hunk 1]
```diff
@@ -181,10 +182,6 @@ public void openConnection(@Nullable String code, @Nullable String redirectUri)
         }
 
         updateStatus(ThingStatus.ONLINE);
-
-        getThing().getThings().stream().filter(Thing::isEnabled).map(Thing::getHandler)
-                .filter(CommonInterface.class::isInstance).map(CommonInterface.class::cast)
-                .forEach(CommonInterface::expireData);
```
Comment (lolodomo) [inline bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java:187]: No more required ?

**Review by lolodomo — COMMENTED** [review]
2 comments

**Comment by lolodomo** [comment]
> I'm available. The PR onboards some other general enhancement, I can restrict it to what's strictly needed if you prefer.

I see that, but I'm used to it with you ;)

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java (hunk 1 above)
Comment (clinique) [reply bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java:187]: No, during the tests I made, I figured that `updateStatus(ThingStatus.ONLINE);` already trigger a general `expireData`. I've experimented many times and it did not appear to add anything.

**Comment by clinique** [comment]
> > I'm available. The PR onboards some other general enhancement, I can restrict it to what's strictly needed if you prefer.
> 
> I see that, but I'm used to it with you ;)

Yes :-) Netatmo binding is so wide that when I see an opportunity to make some obvious enhancements, I take it. It BTW generated this bug.

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/SecurityApi.java [hunk 2]
```diff
@@ -84,10 +81,12 @@ public List<HomeEvent> getHomeEvents(String homeId, ZonedDateTime freshestEventT
...
         // we have to rewind to the latest event just after freshestEventTime
-        if (events.size() > 0) {
+        if (!events.isEmpty()) {
+            String oldestId = "";
             HomeEvent oldestRetrieved = events.get(events.size() - 1);
-            while (oldestRetrieved.getTime().isAfter(freshestEventTime)) {
-                events.addAll(getEvents(PARAM_HOME_ID, homeId, PARAM_EVENT_ID, oldestRetrieved.getId()));
+            while (oldestRetrieved.getTime().isAfter(freshestEventTime) && !oldestId.equals(oldestRetrieved.getId())) {
+                oldestId = oldestRetrieved.getId();
+                events.addAll(getEvents(PARAM_HOME_ID, homeId, PARAM_EVENT_ID, oldestId, PARAM_SIZE, 300));
```
[the remark this answers is not in the document]
Comment (clinique) [reply bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/SecurityApi.java:89]: You're correct, this is an eventuality. Let me add a dedup.

[the remark this answers is not in the document]
Comment (clinique) [reply bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/api/SecurityApi.java:91]: Done

File: bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java (hunk 1 above)
Comment (lolodomo) [reply bundles/org.openhab.binding.netatmo/src/main/java/org/openhab/binding/netatmo/internal/handler/ApiBridgeHandler.java:187]: Hope you're right

**Review by lolodomo — APPROVED** [review]
LGTM, thank you

**Comment by lolodomo** [comment]
> Will look at it this weekend

@lsiepel : Would you like to have a second look?

**Comment by lolodomo** [comment]
@clinique: is it something we can backport in branch 4.2.x ? I mean it does not depend on changes previously done and not backported to 4.2.x ?

**Comment by clinique** [comment]
> @clinique: is it something we can backport in branch 4.2.x ? I mean it does not depend on changes previously done and not backported to 4.2.x ?

No, changes Can and must be backported orelse 4.2.2 will remain dead

**Review by lsiepel — APPROVED** [review]
Thanks. 

@lolodomo i fully trust your review. Thank you for stepping in, I had a busy day and might not have been on time to merge this before milestone 2. I’ll backcourt this if possible.

**Comment by clinique** [comment]
Thanks to you

**Comment by lsiepel** [comment]
Unfortunately it leads to conflicts, can you create a seperate PR against 4.2.x ?

**Comment by clinique** [comment]
> Unfortunately it leads to conflicts, can you create a seperate PR against 4.2.x ?

Sure but I'm not used to. Can you give me some insights on doing PR against a branch (it does not exist in my clinique/openhab-addons fork).

**Comment by clinique** [comment]
> Unfortunately it leads to conflicts, can you create a seperate PR against 4.2.x ?

@lsiepel : done https://github.com/openhab/openhab-addons/pull/17517

**Pull request merged**